packages/go_router/lib/src/route.dart

```diff
@@ -458,4 +541,672 @@ class ShellRoute extends RouteBase {
+/// [ShellNavigatorContainerBuilder] for more details.
+///
+/// Below is a simple example of how a router configuration with
+/// StatefulShellRoute could be achieved. In this example, a
+/// BottomNavigationBar with two tabs is used, and each of the tabs gets its
+/// own Navigator. A container widget responsible for managing the Navigators
+/// for all route branches will then be passed as the child argument
+/// of the builder function.
+///
+/// ```
+/// final GlobalKey<NavigatorState> _tabANavigatorKey =
+///   GlobalKey<NavigatorState>(debugLabel: 'tabANavigator');
+/// final GlobalKey<NavigatorState> _tabBNavigatorKey =
+///   GlobalKey<NavigatorState>(debugLabel: 'tabBNavigator');
+///
+/// final GoRouter _router = GoRouter(
+///   initialLocation: '/a',
+///   routes: <RouteBase>[
+///     StatefulShellRoute(
+///       builder: (BuildContext context, StatefulShellRouteState state,
+///             Widget child) {
+///         return ScaffoldWithNavBar(shellState: state, body: child);
+///       },
+///       branches: [
+///         /// The first branch, i.e. tab 'A'
+///         StatefulShellBranch(
+///           navigatorKey: _tabANavigatorKey,
+///           routes: <RouteBase>[
+///             GoRoute(
+///               path: '/a',
+///               builder: (BuildContext context, GoRouterState state) =>
+///                   const RootScreen(label: 'A', detailsPath: '/a/details'),
+///               routes: <RouteBase>[
+///                 /// Will cover screen A but not the bottom navigation bar
+///                 GoRoute(
+///                   path: 'details',
+///                   builder: (BuildContext context, GoRouterState state) =>
+///                       const DetailsScreen(label: 'A'),
+///                 ),
+///               ],
+///             ),
+///           ],
+///         ),
+///         /// The second branch, i.e. tab 'B'
+///         StatefulShellBranch(
+///           navigatorKey: _tabBNavigatorKey,
+///           routes: <RouteBase>[
+///             GoRoute(
+///               path: '/b',
+///               builder: (BuildContext context, GoRouterState state) =>
+///                   const RootScreen(label: 'B', detailsPath: '/b/details'),
+///               routes: <RouteBase>[
+///                 /// Will cover screen B but not the bottom navigation bar
+///                 GoRoute(
+///                   path: 'details',
+///                   builder: (BuildContext context, GoRouterState state) =>
+///                       const DetailsScreen(label: 'B'),
+///                 ),
+///               ],
+///             ),
+///           ],
+///         ),
+///       ],
+///     ),
+///   ],
+/// );
+/// ```
+///
+/// See [Stateful Nested Navigation](https://github.com/flutter/packages/blob/main/packages/go_router/example/lib/stateful_shell_route.dart)
+/// for a complete runnable example using StatefulShellRoute.
+class StatefulShellRoute extends ShellRouteBase {
```

Comment:
sorry that I have only found out this late, but since I saw the alternative way to implement statefulShellRoute, I wonder if this name maybe to generic.

I imagine we may create more ShellRoute subclass that may have different way to implement the container.

I think we should probably make it more clear in the name. Can we rename this and relevant class to something like StackedShellRoute

Reply:
I guess that would work, even though StatefulShellRoute opens up for other container implementations. But even so, I think StackedShellRoute would work.

Reply:
yes but that would require people to do use the API in a unusual way, i.e. throwing out the child widget. Since there is certain opinionated logic in this class, we should make sure the name reflects that

Reply:
Or alternatively always required users to create the StatefulNavigationShell (or similar).

Reply:
Sounds good - but please have a look at the last commit and see if it is a step in the right direction.